File: NET_project/Benchmarks/Arcfour.cs
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

namespace NET_project.Benchmarks;

[SkipLocalsInit]
public unsafe struct ArcfourNET : IJob
{
	public uint iterations;
	public int result;

	public void Run() {
		result = Arcfour(iterations);
	}

	private static int Arcfour(uint iterations) {
		const int keyLength = 5;
		const int streamLength = 10;

		// Using span here would be more complex due to the general
		// inability to prove that several accesses are safe due to
		// the index being the value of another buffer i.e. x[y[n]]

		byte* state = (byte*)NativeMemory.AlignedAlloc(256, 8);
		byte* buffer = (byte*)NativeMemory.AlignedAlloc(64, 8);
		byte* key = stackalloc byte[keyLength] {
			0xDB,
			0xB7,
			0x60,
			0xD4,
			0x56,
		};
		byte* stream = stackalloc byte[streamLength] {
			0xEB,
			0x9F,
			0x77,
			0x81,
			0xB7,
			0x34,
			0xCA,
			0x72,
			0xA7,
			0x19,
		};

		int idx = 0;

		for (uint i = 0; i < iterations; i++) {
			idx = KeySetup(state, key, keyLength);
			idx = GenerateStream(state, buffer, streamLength);
		}

		NativeMemory.AlignedFree(state);
		NativeMemory.AlignedFree(buffer);

		return idx;
	}

	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	private static int KeySetup(byte* state, byte* key, int keyLength)
	{
		for (uint i = 0; i < 256; i++) {
			state[i] = (byte)i;
		}

		uint j = 0;

		for (uint i = 0; i < 256; i++) {
			j = (j + state[i] + key[i % keyLength]) % 256;
			byte t = state[i];
			state[i] = state[j];
			state[j] = t;
		}

		return 256;
	}

	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	private static int GenerateStream(byte* state, byte* buffer, int length)
	{
		uint i = 0;
		uint j = 0;

		for (int idx = 0; idx < length; idx++) {
			i = (i + 1) % 256;
			j = (j + state[i]) % 256;
			byte t = state[i];
			state[i] = state[j];
			state[j] = t;
			buffer[idx] = state[(state[i] + state[j]) % 256];
		}

		return (int)i;
	}
}

internal unsafe struct ArcfourGCC : IJob
{
	public uint iterations;
	public int result;
	public string libName;

	public void Run()
	{
		switch (libName)
		{
			case "GCC":
				result = GccNativeBindings.benchmark_arcfour(iterations);
				break;

			case "Clang":
				result = ClangNativeBindings.benchmark_arcfour(iterations);
				break;

			case "MS":
				result = MsNativeBindings.benchmark_arcfour(iterations);
				break;

			case "Rust":
				result = RustNativeBindings.benchmark_arcfour(iterations);
				break;
		}
	}
}

File: rust_lib/src/i32_benchmarks/Arcfour.rs


const keyLength: usize = 5;
const streamLength: usize = 10;

fn benchmark_arcfour_key_setup(state: &mut [u8; 256], key: [u8; keyLength], length: usize) -> i32
{
    let mut i: usize;
    let mut j: usize;
    let mut t: u8;

    i = 0;
    while (i < 256)
    {
        state[i] = i as u8;
        i += 1;
    }

    i = 0;
    j = 0;
    while (i < 256)
    {
        j = (j + state[i] as usize + key[i % length] as usize) % 256;
        t = state[i];
        state[i] = state[j];
        state[j] = t;

        i += 1;
    }

    return i as i32;
}

fn benchmark_arcfour_generate_stream(state: &mut [u8; 256], buffer: &mut [u8; 64], length: usize) -> i32
{
    let mut i: usize = 0;
    let mut j: usize = 0;
    let mut idx: usize = 0;
    let mut t: u8;

    while (idx < length)
    {
        i = (i + 1) % 256;
        j = (j + state[i] as usize) % 256;
        t = state[i];
        state[i] = state[j];
        state[j] = t;
        buffer[idx] = state[(state[i] as usize + state[j] as usize) % 256];

        idx += 1;
    }

    return (i as i32);
}

pub(crate) fn benchmark_arcfour_main(iterations: u32) -> i32
{
    let mut state: [u8; 256] = [0; 256];
    let mut buffer: [u8; 64] = [0; 64];
    let mut key: [u8; keyLength] = [0; keyLength];
    let mut stream: [u8; streamLength] = [0; streamLength];

    key[0] = 0xDB;
    key[1] = 0xB7;
    key[2] = 0x60;
    key[3] = 0xD4;
    key[4] = 0x56;

    stream[0] = 0xEB;
    stream[1] = 0x9F;
    stream[2] = 0x77;
    stream[3] = 0x81;
    stream[4] = 0xB7;
    stream[5] = 0x34;
    stream[6] = 0xCA;
    stream[7] = 0x72;
    stream[8] = 0xA7;
    stream[9] = 0x19;

    let mut idx: i32 = 0;

    let mut i: u32 = 0;
    while (i < iterations)
    {
        idx = benchmark_arcfour_key_setup(&mut state, key, keyLength);
        idx = benchmark_arcfour_generate_stream(&mut state, &mut buffer, streamLength);

        i += 1;
    }

    return idx;
}
